### ISSUE TITLE
Celery Rate Limiting on Queue or Worker?

### ISSUE THREAD

**Issue by kpchand** [celery/celery#5732]
#### Related Issues

- None

#### Possible Duplicates

- None

# Description

The mechanism to enforce rate limiting is not mentioned clearly and fails to accomodate different use cases.

[Celery Docs](http://docs.celeryproject.org/en/latest/userguide/tasks.html#Task.rate_limit) state that:

_Note that this is a per worker instance rate limit, and not a global rate limit. To enforce a global rate limit (e.g., for an API with a maximum number of requests per second), you must restrict to a given queue._

# Suggestions
The above fails to clearly mention the mechanism for rate limiting when using multiple workers and my research on internet reveals that there is widespread confusion.

So suppose that one has the 20 reqs/minute rate limit. In such a case to enforce this rate limit, the following can be implemented:

1) Use only 1 worker
2) Use n workers. Divide rate limit by total workers so for 4 workers we have 20/4=5, therefore, the rate limit on task should be set as 4 reqs/min
3) Use task routing to route the rate limited tasks to a unique queue. Using multiple workers will now respect the rate limit as it is performed at queue level.

Now the 3rd point is what is unclear. I request someone with the knowledge to clarify the matter or add it in the docs as the current information is vague.

**Comment by thedrow** [comment]
This will be available in Celery 5.

**Comment by martinburchell** [comment]
Hello. I'm one of the confused people on the internet. How does this feature request differ from what is provided by http://docs.celeryproject.org/en/latest/userguide/workers.html#changing-rate-limits-at-run-time

**Comment by thedrow** [comment]
The global rate limit will be automatically managed by Celery.

**Comment by nicholasamorim** [comment]
Will this be released on Celery 5.0? Can't find any reference in the docs or in the code.

**Comment by thedrow** [comment]
Unfortunately, no. This won't make it to Celery 5.

**Comment by joekohlsdorf** [comment]
@thedrow Can you share the architecture proposal (or the basic idea) for distributed rate limiting?

We are looking into implementing this feature, it would be helpful to get your input if any preliminary work has already been done so that the PR doesn't end up getting delayed unnecessarily.

**Comment by thedrow** [comment]
Hi, we're a long way from that feature.
I'll ping you in a few days with some news on the matter.

**Comment by auvipy** [comment]
> Hi, we're a long way from that feature.
> I'll ping you in a few days with some news on the matter.

is that on celery high level arch cep?

**Comment by mlissner** [comment]
I've been researching this today and realizing that I may have to implement my own global rate limit system. There are a few guides on this, but all of them that I've seen so far use infinite retries to make it work, which is a bit of a bummer. 

I'm just curious: This didn't make it into Celery 5 as hoped. Any idea if it'll make it into the next version? If so, I'll probably wait to implement this myself and just use the built-in system. 

Thank you!

**Comment by nicholasamorim** [comment]
Maybe I am not sure where to look for but it would be amazing if the Celery team could share with us what they have in mind for that feature, architecture-wise or so.

**Comment by joekohlsdorf** [comment]
There is some information on the CEP but it is extremely high level and requires a complete change of the architecture.

https://celery-enhancement-proposals.readthedocs.io/en/latest/draft/high-level-architecture.html#controller
https://celery-enhancement-proposals.readthedocs.io/en/latest/draft/celery-5-high-level-architecture.html#rate-limiting
https://celery-enhancement-proposals.readthedocs.io/en/latest/draft/celery-5-high-level-architecture.html?highlight=rate#concurrency-limitations

tldr: Celery shall have a central component (not sure if that is ever going to happen, I surely don't want a single point of failure) which will broadcast a rate limit event to all workers.
There is no information on how the execution rate of tasks is counted.

**Comment by hardbyte** [comment]
I'm curious as to whether point 3 is a valid approach for now?

> Use task routing to route the rate limited tasks to a unique queue. Using multiple workers will now respect the rate limit as it is performed at queue level.

**Comment by thedrow** [comment]
> tldr: Celery shall have a central component (not sure if that is ever going to happen, I surely don't want a single point of failure) which will broadcast a rate limit event to all workers.
> There is no information on how the execution rate of tasks is counted.

It's not a single point of failure since it is highly available.

**Comment by NoraGithub** [comment]
is it  available  now?

**Comment by radiantone** [comment]
You can implement throttling by sleeping for some period of time in the @task_prerun.connect handler. I do this and it works just fine. The celery internals will block and provide backpressure to upstream invocations. It's not "queue based" per se, but it's an easy and effective stop-gap

**Comment by joekohlsdorf** [comment]
> You can implement throttling by sleeping for some period of time in the @task_prerun.connect handler. I do this and it works just fine. The celery internals will block and provide backpressure to upstream invocations. It's not "queue based" per se, but it's an easy and effective stop-gap

This is a horrible recommendation unless you only process a single task in your workers because you'll just block workers from processing other tasks by actively sleeping.

**Comment by man4red** [comment]
There's a Token Bucket approach as for example 
source https://habr.com/ru/articles/494090/ 

```python
from celery import Celery
from kombu import Queue
from queue import Empty
from functools import wraps

app = Celery('hello', broker='amqp://guest@localhost//')

task_queues = [
    Queue('github'),
    Queue('google')
]

# rate per minute
rate_limits = {
    'github': 60,
    'google': 100
}

# create token queues
task_queues += [Queue(name+'_tokens', max_length=2) for name, limit in rate_limits.items()]

app.conf.task_queues = task_queues

# token issuance 
@app.task
def token():
    return 1

@app.on_after_configure.connect
def setup_periodic_tasks(sender, **kwargs):
    for name, limit in rate_limits.items():
        sender.add_periodic_task(60 / limit, token.signature(queue=name + "_tokens"), name=name + "_tokens")

def rate_limit(task_group):
    def decorator_func(func):
        @wraps(func)
        def function(self, *args, **kwargs):
            with self.app.connection_for_read() as conn:
                with conn.SimpleQueue(task_group+'_tokens', no_ack=True, queue_opts={'max_length':2}) as queue:
                    try:
                        queue.get(block=True, timeout=10)
                        return func(self, *args, **kwargs)
                    except Empty:
                        self.retry(countdown=1)
        return function
    return decorator_func

@app.task(bind=True, max_retries=None)
@rate_limit('github')
def get_github_api1(self):
    print ('Called github Api 1')

@app.task(bind=True, max_retries=None)
@rate_limit('github')
def get_github_api2(self):
    print ('Called github Api 2')

@app.task(bind=True, max_retries=None)
@rate_limit('google')
def query_google_api1(self):
    print ('Called Google Api 1')

@app.task(bind=True, max_retries=None)
@rate_limit('google')
def query_google_api1(self):
    print ('Called Google Api 2')
```

**Comment by joekohlsdorf** [comment]
This doesn't work at scale, you end up in a busy loop between your queue (Redis/RabbitMQ) and Celery workers.
The countdown=1 won't help much if you have a large number of queued tasks.

**Comment by man4red** [comment]
> This doesn't work at scale, you end up in a busy loop between your queue (Redis/RabbitMQ) and Celery workers. The countdown=1 won't help much if you have a large number of queued tasks.

It's a proof of concept, I guess. The Token Bucket approach looks like a good solution in general. You mentioned a 'centralized solution,' and I completely agree that it could be a single point of failure.

Currently, I'm using beat with locks (redbeat.RedBeatScheduler), allowing me to run multiple beats (although only one operates at a time). What are your thoughts on having a single (and the only one by utilizing locks) Celery worker instance issue tokens per queue? This way, tasks could utilize them as a global rate limiter?
